clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -212,9 +212,9 @@ public void connect(String id, InetSocketAddress address, int sendBufferSize, in
         try {
             configureSocketChannel(socketChannel, sendBufferSize, receiveBufferSize);
             boolean connected = doConnect(socketChannel, address);
-            SelectionKey key = registerChannel(id, socketChannel, SelectionKey.OP_CONNECT);
 
             if (connected) {
+                SelectionKey key = registerChannel(id, socketChannel, SelectionKey.OP_CONNECT);
```

Comment:
I don't think this works. This branch only handles connections which are completed "immediately" in `doConnect`. This is not the common case, which is why all of the test cases in `SelectorTest` fail with this change.